jbehave-core/src/main/java/org/jbehave/core/steps/StepCreator.java

```diff
@@ -824,6 +835,91 @@ public StepResult perform(StoryReporter storyReporter, UUIDExceptionWrapper stor
+
+        public ConditionalStep(StepConditionMatcher stepConditionMatcher,
+                Map<Method, ParametrisedStep> parametrisedSteps) {
+            this.stepConditionMatcher = stepConditionMatcher;
+            this.parametrisedSteps = parametrisedSteps;
+        }
+
+        @Override
+        public StepResult perform(StoryReporter storyReporter, UUIDExceptionWrapper storyFailureIfItHappened) {
+            return performConditionalStep(step -> step.perform(storyReporter, storyFailureIfItHappened), true);
+        }
+
+        @Override
+        public StepResult doNotPerform(StoryReporter storyReporter, UUIDExceptionWrapper storyFailureIfItHappened) {
+            return performConditionalStep(step -> step.doNotPerform(storyReporter, storyFailureIfItHappened), false);
+        }
+
+        private StepResult performConditionalStep(Function<ParametrisedStep, StepResult> invoker,
+                boolean reportToStepMonitor) {
+
+            try {
+                Map<Method, Conditional> unmatchedSteps = new HashMap<>();
+                Map<Method, Conditional> matchedSteps = new HashMap<>();
+
+                for (Method method : parametrisedSteps.keySet()) {
+                    Conditional conditional = Optional.ofNullable(method.getAnnotation(Conditional.class))
+                            .orElseGet(() -> method.getDeclaringClass().getAnnotation(Conditional.class));
+                    if (stepConditionMatcher.matches(conditional.condition(), conditional.value())) {
+                        matchedSteps.put(method, conditional);
+                    } else {
+                        unmatchedSteps.put(method, conditional);
+                    }
+                }
+
+                if (matchedSteps.isEmpty()) {
+                    String message = "None of the following steps were matched any condition:" + System.lineSeparator()
+                            + formatSteps(unmatchedSteps);
+                    return reportPending(message, reportToStepMonitor);
+                } else if (matchedSteps.size() > 1) {
+                    String message = "More than one conditional step matched the condition:" + System.lineSeparator()
+                            + formatSteps(matchedSteps);
+                    return reportPending(message, reportToStepMonitor);
+                }
+
+                return invoker.apply(parametrisedSteps.get(matchedSteps.keySet().iterator().next()));
+            } catch (StepConditionMatchException e) {
+                return reportPending(e.getMessage(), reportToStepMonitor);
+            }
+        }
+
+        private StepResult reportPending(String message, boolean reportToStepMonitor) {
+            String parametrisedStep = parametrisedSteps.values().iterator().next().asString(null);
+            try {
+                if (reportToStepMonitor) {
+                    stepMonitor.beforePerforming(parametrisedStep, dryRun, null);
```

Comment:
it should be invoked for `ParametrisedStep` only, but for `PendingStep` it's needed to invoke `org.jbehave.core.reporters.StoryReporter#beforeStep`